**INvestigation To Perform**

- 1. PostgreSQL Connection
- 2. Staging Validation
- 3. Dataset Overview
- 4. Cardinality Analysis
- 5. Functional Dependency Analysis
- 6. Repeating-Group Analysis
- 7. Location Analysis
- 8. Vehicle Analysis
- 9. Contributing-Factor Analysis
- 10. Normalization Evidence
- 11. Design Conclusions

The findings will be used to design the normalized 3NF model.

## Architecture

Cleaned CSV                                                                     
    ↓
PostgreSQL Staging                                                              
    ↓
SQLAlchemy                                                                  
    ↓
SQL Investigation                                               
    ↓
Pandas                                                                          
    ↓
Analysis / Visualization                                                        
    ↓
3NF Design                                                          

In [ ]:
import pandas as pd
import numpy as np
# ! pip install sqlalchemy
# ! pip install psycopg2-binary
from sqlalchemy import create_engine, text

import matplotlib.pyplot as plt
import plotly.express as px

In [5]:
# Creating engine to connect Database from PostgreSQL to Python
engine = create_engine(
    "postgresql+psycopg2://postgres:postgres@localhost:5432/nyc_collisions"
)

In [8]:
query = """
SELECT version();
"""

version_df = pd.read_sql(query, engine)

version_df

,version
0,PostgreSQL 16.4 (Debian 16.4-1.pgdg110+2) on x...


In [9]:
query = """
SELECT PostGIS_Version();
"""

postgis_df = pd.read_sql(query, engine)

postgis_df

,postgis_version
0,3.4 USE_GEOS=1 USE_PROJ=1 USE_STATS=1


### Investigation 

**1. Row count**

In [10]:
query = """
SELECT COUNT(*) AS total_rows
FROM staging.cleaned_collisions;
"""

row_count = pd.read_sql(query, engine)

row_count

,total_rows
0,2028275


**2. Columns Count**

In [11]:
query = """
SELECT COUNT(*) AS total_columns
FROM information_schema.columns
WHERE table_schema = 'staging'
  AND table_name = 'cleaned_collisions';
"""

column_count = pd.read_sql(query, engine)

column_count

,total_columns
0,28


**3. Actual Database Schema**

In [12]:
query = """
SELECT
    ordinal_position,
    column_name,
    data_type
FROM information_schema.columns
WHERE table_schema = 'staging'
  AND table_name = 'cleaned_collisions'
ORDER BY ordinal_position;
"""

schema_df = pd.read_sql(query, engine)

schema_df

,ordinal_position,column_name,data_type
0,1,crash_date,date
1,2,crash_time,time without time zone
2,3,borough,text
3,4,zip_code,text
4,5,latitude,numeric
5,6,longitude,numeric
6,7,on_street_name,text
7,8,cross_street_name,text
8,9,off_street_name,text
9,10,number_of_persons_injured,integer


### Candidate Key Investigation (COLLISION_ID)

***1. Does COLLISION_ID uniquely identify a collision record in the cleaned staging data?***

In [13]:
query = """
SELECT
    COUNT(*) AS total_rows,
    COUNT(collision_id) AS non_null_collision_ids,
    COUNT(DISTINCT collision_id) AS distinct_collision_ids
FROM staging.cleaned_collisions;
"""

collision_id_summary = pd.read_sql(query, engine)

collision_id_summary

,total_rows,non_null_collision_ids,distinct_collision_ids
0,2028275,2028275,2028275


In [14]:
query = """
SELECT
    collision_id,
    COUNT(*) AS record_count
FROM staging.cleaned_collisions
GROUP BY collision_id
HAVING COUNT(*) > 1
ORDER BY record_count DESC
LIMIT 20;
"""

collision_duplicates = pd.read_sql(query, engine)

collision_duplicates

,collision_id,record_count


In [15]:
print("Number of duplicate collision IDs:", len(collision_duplicates))

Number of duplicate collision IDs: 0


**Finding — COLLISION_ID**

The staging table contains 2,028,275 collision records.

`COLLISION_ID` is non-null for all records and the number of distinct
`COLLISION_ID` values equals the total number of records.

No duplicate `COLLISION_ID` values were identified.

Therefore, `COLLISION_ID` is a valid candidate key for the collision
entity in the normalized model.

This supports the dependency:

`COLLISION_ID → collision-level attributes`

### **Location Investigation**

**A. Coordinate cardinality**

In [16]:
query = """
SELECT
    COUNT(*) AS total_rows,
    COUNT(*) FILTER (
        WHERE latitude IS NOT NULL
          AND longitude IS NOT NULL
    ) AS rows_with_coordinates,
    COUNT(DISTINCT (latitude, longitude)) AS distinct_coordinates
FROM staging.cleaned_collisions;
"""

coordinate_summary = pd.read_sql(query, engine)

coordinate_summary

,total_rows,rows_with_coordinates,distinct_coordinates
0,2028275,2028275,348400


In [30]:
query = """
SELECT
    latitude,
    longitude,
    COUNT(*) AS collision_count
FROM staging.cleaned_collisions
WHERE latitude IS NOT NULL
  AND longitude IS NOT NULL
GROUP BY latitude, longitude
ORDER BY collision_count DESC;
"""

coordinate_frequency = pd.read_sql(query, engine)

coordinate_frequency.head(20)

,latitude,longitude,collision_count
0,0.000000,0.000000,7591
1,40.608757,-74.038086,887
2,40.696033,-73.984530,712
3,40.861862,-73.912820,685
4,40.804700,-73.912430,597
5,40.696035,-73.984529,587
6,40.675735,-73.896860,581
7,40.658577,-73.890630,527
8,40.604153,-74.051980,510
9,40.798256,-73.827440,477


In [31]:
query = """
SELECT
    MIN(latitude) AS min_latitude,
    MAX(latitude) AS max_latitude,
    MIN(longitude) AS min_longitude,
    MAX(longitude) AS max_longitude
FROM staging.cleaned_collisions
WHERE latitude IS NOT NULL
  AND longitude IS NOT NULL
  AND NOT (latitude = 0 AND longitude = 0);
"""

coordinate_range_df = pd.read_sql(query, engine)

coordinate_range_df

,min_latitude,max_latitude,min_longitude,max_longitude
0,30.78418,43.344444,-89.13527,-32.768513


### Finding — Invalid Zero Coordinates

The investigation identified 7,591 records with latitude and longitude
equal to `(0, 0)`.

Because `(0, 0)` does not represent a valid NYC geographic location,
these records should not be treated as valid spatial observations.

The records will remain in the staging layer as source data. During
the PostGIS transformation, coordinate validation will be applied and
invalid coordinates will not be converted into spatial geometry.

Valid latitude/longitude pairs will be treated as the authoritative
spatial pointer for the collision.

### **Investigation 3 — Vehicle Repeating Groups**

**1. First: quantify vehicle-slot usage**

In [32]:
query = """
SELECT
    COUNT(*) FILTER (
        WHERE vehicle_type_code_1 IS NOT NULL
    ) AS vehicle_1_count,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_2 IS NOT NULL
    ) AS vehicle_2_count,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_3 IS NOT NULL
    ) AS vehicle_3_count,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_4 IS NOT NULL
    ) AS vehicle_4_count,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_5 IS NOT NULL
    ) AS vehicle_5_count
FROM staging.cleaned_collisions;
"""

vehicle_slot_summary = pd.read_sql(query, engine)

vehicle_slot_summary

,vehicle_1_count,vehicle_2_count,vehicle_3_count,vehicle_4_count,vehicle_5_count
0,2012458,1601962,140364,32611,9129


In [33]:
vehicle_slots = vehicle_slot_summary.T.reset_index()

vehicle_slots.columns = ["vehicle_slot", "records"]

fig = px.bar(
    vehicle_slots,
    x="vehicle_slot",
    y="records",
    title="Vehicle Type Slot Utilization",
    labels={
        "vehicle_slot": "Vehicle Type Slot",
        "records": "Records"
    }
)

fig.show()

**3. How many distinct vehicle types exist?**

In [34]:
query = """
SELECT
    vehicle_type,
    COUNT(*) AS occurrences
FROM (
    SELECT vehicle_type_code_1 AS vehicle_type
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT vehicle_type_code_2
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT vehicle_type_code_3
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT vehicle_type_code_4
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT vehicle_type_code_5
    FROM staging.cleaned_collisions
) AS vehicles
WHERE vehicle_type IS NOT NULL
GROUP BY vehicle_type
ORDER BY occurrences DESC;
"""

vehicle_types = pd.read_sql(query, engine)

vehicle_types.head(20)

,vehicle_type,occurrences
0,SEDAN,1094299
1,STATION WAGON/SPORT UTILITY VEHICLE,862705
2,PASSENGER VEHICLE,640442
3,SPORT UTILITY / STATION WAGON,282299
4,TAXI,148244
5,PICK-UP TRUCK,90663
6,UNKNOWN,90600
7,BUS,67874
8,VAN,62619
9,4 DR SEDAN,57967


In [41]:
fig = px.bar(
    vehicle_types.head(20),
    x="vehicle_type",
    y="occurrences",
    title="Top 20 Vehicle Types",
    labels={
        "occurrences": "Occurrences",
        "vehicle_type": "Vehicle Type"
    }
)

fig.show()

**4. Number of Collisions where same vehicle type occurred**

In [42]:
query = """
SELECT COUNT(*) AS collisions_with_repeated_vehicle_type
FROM staging.cleaned_collisions
WHERE
    (vehicle_type_code_1 IS NOT NULL AND vehicle_type_code_1 = vehicle_type_code_2)
 OR (vehicle_type_code_1 IS NOT NULL AND vehicle_type_code_1 = vehicle_type_code_3)
 OR (vehicle_type_code_1 IS NOT NULL AND vehicle_type_code_1 = vehicle_type_code_4)
 OR (vehicle_type_code_1 IS NOT NULL AND vehicle_type_code_1 = vehicle_type_code_5)
 OR (vehicle_type_code_2 IS NOT NULL AND vehicle_type_code_2 = vehicle_type_code_3)
 OR (vehicle_type_code_2 IS NOT NULL AND vehicle_type_code_2 = vehicle_type_code_4)
 OR (vehicle_type_code_2 IS NOT NULL AND vehicle_type_code_2 = vehicle_type_code_5)
 OR (vehicle_type_code_3 IS NOT NULL AND vehicle_type_code_3 = vehicle_type_code_4)
 OR (vehicle_type_code_3 IS NOT NULL AND vehicle_type_code_3 = vehicle_type_code_5)
 OR (vehicle_type_code_4 IS NOT NULL AND vehicle_type_code_4 = vehicle_type_code_5);
"""

repeated_vehicle_types = pd.read_sql(query, engine)

repeated_vehicle_types

,collisions_with_repeated_vehicle_type
0,659765


### Finding — Vehicle Repeating Group

The source data stores vehicle types across five repeating columns:

- `VEHICLE TYPE CODE 1`
- `VEHICLE TYPE CODE 2`
- `VEHICLE TYPE CODE 3`
- `VEHICLE TYPE CODE 4`
- `VEHICLE TYPE CODE 5`

The investigation identified 659,765 collision records in which the same
vehicle type appears in more than one vehicle position.

Therefore, vehicle occurrences cannot be represented as a unique set of
vehicle types per collision.

The normalized model should represent vehicle involvement as a 1:N
relationship between a collision and its vehicle occurrences, while
preserving the original vehicle position.

Proposed structure:

`Collision → CollisionVehicle → VehicleType`

### **Investigation 4 Contributing-Factor Investigation**

In [43]:
# Factor Slot Utilization
query = """
SELECT
    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_1 IS NOT NULL
    ) AS factor_1_count,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_2 IS NOT NULL
    ) AS factor_2_count,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_3 IS NOT NULL
    ) AS factor_3_count,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_4 IS NOT NULL
    ) AS factor_4_count,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_5 IS NOT NULL
    ) AS factor_5_count
FROM staging.cleaned_collisions;
"""

factor_slot_summary = pd.read_sql(query, engine)

factor_slot_summary

,factor_1_count,factor_2_count,factor_3_count,factor_4_count,factor_5_count
0,2020530,1691725,146336,33869,9419


In [44]:
# Distinct Factor Values
query = """
SELECT
    contributing_factor,
    COUNT(*) AS occurrences
FROM (
    SELECT contributing_factor_vehicle_1 AS contributing_factor
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT contributing_factor_vehicle_2
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT contributing_factor_vehicle_3
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT contributing_factor_vehicle_4
    FROM staging.cleaned_collisions

    UNION ALL

    SELECT contributing_factor_vehicle_5
    FROM staging.cleaned_collisions
) AS factors
WHERE contributing_factor IS NOT NULL
GROUP BY contributing_factor
ORDER BY occurrences DESC;
"""

factors = pd.read_sql(query, engine)

factors.head(20)

,contributing_factor,occurrences
0,UNSPECIFIED,2274102
1,DRIVER INATTENTION/DISTRACTION,514465
2,FAILURE TO YIELD RIGHT-OF-WAY,141456
3,FOLLOWING TOO CLOSELY,125851
4,OTHER VEHICULAR,99030
5,BACKING UNSAFELY,83445
6,PASSING OR LANE USAGE IMPROPER,72456
7,PASSING TOO CLOSELY,63626
8,TURNING IMPROPERLY,56773
9,FATIGUED/DROWSY,47305


In [45]:
# Repeated Factors For Collision
query = """
SELECT
    COUNT(*) AS collisions_with_repeated_factor
FROM staging.cleaned_collisions
WHERE
    (
        contributing_factor_vehicle_1 IS NOT NULL
        AND contributing_factor_vehicle_1 = contributing_factor_vehicle_2
    )
    OR
    (
        contributing_factor_vehicle_1 IS NOT NULL
        AND contributing_factor_vehicle_1 = contributing_factor_vehicle_3
    )
    OR
    (
        contributing_factor_vehicle_1 IS NOT NULL
        AND contributing_factor_vehicle_1 = contributing_factor_vehicle_4
    )
    OR
    (
        contributing_factor_vehicle_1 IS NOT NULL
        AND contributing_factor_vehicle_1 = contributing_factor_vehicle_5
    )
    OR
    (
        contributing_factor_vehicle_2 IS NOT NULL
        AND contributing_factor_vehicle_2 = contributing_factor_vehicle_3
    )
    OR
    (
        contributing_factor_vehicle_2 IS NOT NULL
        AND contributing_factor_vehicle_2 = contributing_factor_vehicle_4
    )
    OR
    (
        contributing_factor_vehicle_2 IS NOT NULL
        AND contributing_factor_vehicle_2 = contributing_factor_vehicle_5
    )
    OR
    (
        contributing_factor_vehicle_3 IS NOT NULL
        AND contributing_factor_vehicle_3 = contributing_factor_vehicle_4
    )
    OR
    (
        contributing_factor_vehicle_3 IS NOT NULL
        AND contributing_factor_vehicle_3 = contributing_factor_vehicle_5
    )
    OR
    (
        contributing_factor_vehicle_4 IS NOT NULL
        AND contributing_factor_vehicle_4 = contributing_factor_vehicle_5
    );
"""

repeated_factor_df = pd.read_sql(query, engine)

repeated_factor_df

,collisions_with_repeated_factor
0,716966


In [46]:
query = """
SELECT
    COUNT(*) FILTER (
        WHERE vehicle_type_code_1 IS NOT NULL
          AND contributing_factor_vehicle_1 IS NOT NULL
    ) AS vehicle1_factor1_both,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_1 IS NOT NULL
          AND contributing_factor_vehicle_1 IS NULL
    ) AS vehicle1_without_factor1,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_2 IS NOT NULL
          AND contributing_factor_vehicle_2 IS NOT NULL
    ) AS vehicle2_factor2_both,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_2 IS NOT NULL
          AND contributing_factor_vehicle_2 IS NULL
    ) AS vehicle2_without_factor2,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_3 IS NOT NULL
          AND contributing_factor_vehicle_3 IS NOT NULL
    ) AS vehicle3_factor3_both,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_4 IS NOT NULL
          AND contributing_factor_vehicle_4 IS NOT NULL
    ) AS vehicle4_factor4_both,

    COUNT(*) FILTER (
        WHERE vehicle_type_code_5 IS NOT NULL
          AND contributing_factor_vehicle_5 IS NOT NULL
    ) AS vehicle5_factor5_both

FROM staging.cleaned_collisions;
"""

factor_vehicle_alignment = pd.read_sql(query, engine)

factor_vehicle_alignment

,vehicle1_factor1_both,vehicle1_without_factor1,vehicle2_factor2_both,vehicle2_without_factor2,vehicle3_factor3_both,vehicle4_factor4_both,vehicle5_factor5_both
0,2009596,2862,1576545,25417,138942,32294,9036


In [47]:
query = """
SELECT
    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_1 IS NOT NULL
          AND vehicle_type_code_1 IS NULL
    ) AS factor1_without_vehicle1,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_2 IS NOT NULL
          AND vehicle_type_code_2 IS NULL
    ) AS factor2_without_vehicle2,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_3 IS NOT NULL
          AND vehicle_type_code_3 IS NULL
    ) AS factor3_without_vehicle3,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_4 IS NOT NULL
          AND vehicle_type_code_4 IS NULL
    ) AS factor4_without_vehicle4,

    COUNT(*) FILTER (
        WHERE contributing_factor_vehicle_5 IS NOT NULL
          AND vehicle_type_code_5 IS NULL
    ) AS factor5_without_vehicle5

FROM staging.cleaned_collisions;
"""

factor_without_vehicle = pd.read_sql(query, engine)

factor_without_vehicle

,factor1_without_vehicle1,factor2_without_vehicle2,factor3_without_vehicle3,factor4_without_vehicle4,factor5_without_vehicle5
0,10934,115180,7394,1575,383


In [48]:
query = """
SELECT
    COUNT(*) AS collisions_with_repeated_factor
FROM staging.cleaned_collisions
WHERE
    (contributing_factor_vehicle_1 IS NOT NULL
     AND contributing_factor_vehicle_1 = contributing_factor_vehicle_2)
 OR (contributing_factor_vehicle_1 IS NOT NULL
     AND contributing_factor_vehicle_1 = contributing_factor_vehicle_3)
 OR (contributing_factor_vehicle_1 IS NOT NULL
     AND contributing_factor_vehicle_1 = contributing_factor_vehicle_4)
 OR (contributing_factor_vehicle_1 IS NOT NULL
     AND contributing_factor_vehicle_1 = contributing_factor_vehicle_5)
 OR (contributing_factor_vehicle_2 IS NOT NULL
     AND contributing_factor_vehicle_2 = contributing_factor_vehicle_3)
 OR (contributing_factor_vehicle_2 IS NOT NULL
     AND contributing_factor_vehicle_2 = contributing_factor_vehicle_4)
 OR (contributing_factor_vehicle_2 IS NOT NULL
     AND contributing_factor_vehicle_2 = contributing_factor_vehicle_5)
 OR (contributing_factor_vehicle_3 IS NOT NULL
     AND contributing_factor_vehicle_3 = contributing_factor_vehicle_4)
 OR (contributing_factor_vehicle_3 IS NOT NULL
     AND contributing_factor_vehicle_3 = contributing_factor_vehicle_5)
 OR (contributing_factor_vehicle_4 IS NOT NULL
     AND contributing_factor_vehicle_4 = contributing_factor_vehicle_5);
"""

repeated_factor_df = pd.read_sql(query, engine)

repeated_factor_df

,collisions_with_repeated_factor
0,716966


### **Investigation 5 - Location Investigation**

**Check 1 — Location cardinality**

In [49]:
query = """
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT (
        latitude,
        longitude,
        borough,
        zip_code,
        on_street_name,
        cross_street_name,
        off_street_name
    )) AS distinct_location_combinations
FROM staging.cleaned_collisions;
"""

location_cardinality = pd.read_sql(query, engine)

location_cardinality

,total_rows,distinct_location_combinations
0,2028275,499825


**Check 2 — Collision grain**

In [50]:
query = """
SELECT
    COUNT(*) AS staging_rows,
    COUNT(DISTINCT collision_id) AS distinct_collisions
FROM staging.cleaned_collisions;
"""

grain_check = pd.read_sql(query, engine)

grain_check

,staging_rows,distinct_collisions
0,2028275,2028275


### Final 3NF Model After Investigation
                         ┌──────────────────────┐
                         │ normalized.collisions│
                         │                      │
                         │ collision_id PK      │
                         │ crash_date           │
                         │ crash_time           │
                         │ location attributes  │
                         │ collision measures   │
                         └──────────┬───────────┘
                                    │
                       ┌────────────┴─────────────┐
                       │                          │
                      1:N                        1:N
                       │                          │
                       ▼                          ▼
          ┌──────────────────────┐   ┌──────────────────────┐
          │ collision_vehicles   │   │ collision_factors    │
          │                      │   │                      │
          │ collision_id         │   │ collision_id         │
          │ vehicle_sequence     │   │ factor_sequence      │
          │ vehicle_type_key     │   │ factor_key           │
          └──────────┬───────────┘   └──────────┬───────────┘
                     │                          │
                     ▼                          ▼
          ┌──────────────────┐       ┌─────────────────────┐
          │ vehicle_types    │       │ contributing_factors│
          │                  │       │                     │
          │ vehicle_type_key │       │ factor_key         │
          │ vehicle_type     │       │ factor_description │
          └──────────────────┘       └─────────────────────┘

# Final Investigation Conclusions

The investigation established the following database design decisions:

1. The staging table has a grain of one row per collision event.

2. `COLLISION_ID` uniquely identifies each collision and is therefore the
   natural/business key for the collision entity.

3. Vehicle type attributes form a repeating group across five source
   columns. Vehicle occurrences will therefore be represented in a
   separate `collision_vehicles` table with a preserved vehicle sequence.

4. Contributing factors form a repeating group across five source columns.
   Factors will therefore be represented in a separate `collision_factors`
   table with a preserved factor sequence.

5. Vehicle and contributing-factor sequences are not treated as a strict
   one-to-one relationship because contributing factors can exist without
   a corresponding vehicle slot.

6. Repeated vehicle types and repeated contributing factors within the
   same collision are preserved as separate occurrences rather than
   deduplicated.

7. Latitude and longitude are retained as the authoritative spatial
   coordinates for later PostGIS analysis.

8. Coordinate pairs are not assumed to functionally determine borough or
   ZIP code because the investigation identified conflicting administrative
   attributes for identical coordinates.

9. Location attributes will remain associated with the collision in the
   normalized layer. A reusable `DimLocation` will be introduced later
   for the dimensional warehouse and spatial analysis.

10. Derived temporal attributes will be generated later in `DimDate`
    rather than retained in the staging representation.

The normalized schema can now be implemented in PostgreSQL.